Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -245,26 +233,75 @@ private Task<string> FetchAppIdFromService(string instrumentationKey)
                 {
                     SdkInternalOperationsMonitor.Enter();
 
+                    string result = null;
                     Uri appIdEndpoint = this.GetAppIdEndPointUri(instrumentationKey);
-
                     WebRequest request = WebRequest.Create(appIdEndpoint);
                     request.Method = "GET";
 
                     using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
                     {
-                        using (StreamReader reader = new StreamReader(response.GetResponseStream()))
+                        if (response.StatusCode == HttpStatusCode.OK)
+                        {
+                            using (StreamReader reader = new StreamReader(response.GetResponseStream()))
+                            {
+                                result = reader.ReadToEnd();
+                            }
+                        }
+                        else
                         {
-                            return reader.ReadToEnd();
+                            this.RegisterFetchFailure(instrumentationKey, response.StatusCode);
                         }
                     }
+
+                    return result;
                 }
                 finally
                 {
                     SdkInternalOperationsMonitor.Exit();
                 }
             });
         }
+#else
+        /// <summary>
+        /// Retrieves the app id given the instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key for which app id is to be retrieved.</param>
+        /// <returns>App id.</returns>
+        private async Task<string> FetchAppIdFromService(string instrumentationKey)
+        {
+            try
+            {
+                SdkInternalOperationsMonitor.Enter();
+
+                string result = null;
+                Uri appIdEndpoint = this.GetAppIdEndPointUri(instrumentationKey);
+                WebRequest request = WebRequest.Create(appIdEndpoint);
+                request.Method = "GET";
+
+                using (HttpWebResponse response = (HttpWebResponse)await request.GetResponseAsync().ConfigureAwait(false))
+                {
+                    if (response.StatusCode == HttpStatusCode.OK)
+                    {
+                        using (StreamReader reader = new StreamReader(response.GetResponseStream()))
+                        {
+                            result = await reader.ReadToEndAsync();
```

Comment:
Since the issue is a bad response - should we limit the size of response and maybe filter out special characters and such?

Reply:
The original bug was an error to put a non-ascii LineReturn character in an http header.
Assuming we trust our services, we shouldn't have to sanitize a successful response.
And these changes will not parse the response body of any failed requests, so I think we're covered.

If you disagree we can scrub it.

Reply:
even if we trust our server we still need to make sure that return string will not break customer application. So there should be a sanitization logic either here or when we serialize app-id into the http header.

Reply:
That makes sense. I added a sanitization check (line 195: Guid.TryParse()) after the fact when we add the app-id to the dictionary. I chose to do it here because the FetchAppIdFromService() method can be overridden at runtime.

Reply:
I'd make it less format-specific and more http-safe. Just strip out all characters forbidden in http header value. This way we can encode something into `app-id` in future if we will need to. Like prefix or something